BetterSubsystemBase & DriverUtils & MotorUtils

LaurensDSchunk opened the pull request:
Imported BetterSubsystemBase, DriverUtils, and MotorUtils.

For those who don't know:
 - BetterSubsystemBase is a derived from SubsystemBase and adds motor logging and faster PID configuration. 🔥
 - DriverUtils is for alerting the driver when 45 seconds left using controller vibration. 🫨
 - MotorUtils (required by BetterSubsystemBase) provides a PID and Motor struct for configuring and debugging motors. 🪲

Cynosure-null commented on src/main/cpp/utility/BetterSubsystemBase.cpp:
```diff
@@ -0,0 +1,44 @@
+void BetterSubsystemBase::AddPID(std::vector<MotorUtils::Motor> motors) {
+  this->motors = motors;
+
+  for (int i = 0; i < this->motors.size(); i++) {
```
You can use `for (auto &i: this->motors)` instead for brevity and readability. You don't need to change this but it's just good to know.

Cynosure-null commented on src/main/cpp/utility/BetterSubsystemBase.cpp:
```diff
@@ -0,0 +1,44 @@
+    configs.Slot0.kD = PIDValue.kD; 
+
+    ctre::phoenix::StatusCode status = ctre::phoenix::StatusCode::StatusCodeNotInitialized;
+    for (int j = 0; j < 5; j++)
```
Why is 5 hardcoded here?

Cynosure-null commented on src/main/cpp/utility/BetterSubsystemBase.cpp:
```diff
@@ -0,0 +1,44 @@
+    }
+    if (!status.IsOK())
+    {
+        std::cout << "Could not apply configs, error code: " << status.GetName() << std::endl;
```
Use `frc::DataLogManager::Log()` instead of cout

Cynosure-null commented on src/main/cpp/utility/DriverUtils.cpp:
```diff
@@ -0,0 +1,16 @@
+  // double result = units::unit_cast<double>(time);
+  frc::SmartDashboard::PutNumber("time", time);
+  int duration = 1;
+  if (time <= 45 && time > 45 - duration) {
```
Why are you comparing `double time` and `int duration`?

Cynosure-null commented on src/main/include/utility/MotorUtils.h:
```diff
@@ -0,0 +1,37 @@
+    };
+
+    ctre::phoenix6::hardware::TalonFX *motorPtr;
+    std::string name;
```
Consider setting name to the name given to the device in TunerX via `.GetDescription()`

Cynosure-null reviewed (changes requested):
Additionally, can you link an example of this being used for documentation?

LaurensDSchunk commented:
I changed the naming to use the motor's device name. Replaced my index-based for-loops with advanced for loops. `PidCoeff` struct moved from constants to motorUtils for better clarity.
